New vmath functions

ekharkunov opened the pull request:
Added new `vmath` functions:
* `vmath.quat_matrix4` - get quaternion from matrix4
* `vmath.matrix4_compose` - create matrix4 from translation (vector3|vector4), rotation (quaternion) and scale (vector3)
* `vmath.matrix4_scale` - create matrix4 from scale (vector3) or from number (uniform scale) or 3 number (x, y, z)
* `vmath.clamp` - clamp input value in range [min, max]. Input value can be number, vector, vector3, vector4. In case of vector type clamp applies per vector component. `min`, `max` also can be number, vector, vector3, vector4.

Fixes #5118 

## PR checklist

* [x] Code
	* [x] Add engine and/or editor unit tests.
	* [x] New and changed code follows the overall code style of existing code
	* [x] Add comments where needed
* [x] Documentation
	* [x] Make sure that API documentation is updated in code comments
	* [x] Make sure that manuals are updated (in github.com/defold/doc)
* [x] Prepare pull request and affected issue for automatic release notes generator
	* [x] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [x] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [x] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [x] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [x] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [x] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

ekharkunov committed:
New vmath functions.

JCash commented on engine/script/src/script_vmath.cpp:
````diff
@@ -2299,6 +2299,166 @@ namespace dmScript
+     * print(result) --> vmath.matrix4(-25, -10, 11, 103, 28, -9.5, 2, -95, -10, 10, -4.5, 14, 0, 0, 0, 1)
+     * ```
+     */
+    static int Matrix4_Compose(lua_State* L)
````
Potentially, it could be part of the constructor as well.
I.e. a `vmath.matrix4(t, r, s)` (and maybe a `vmath.matrix4(m00, m01, ...)`

Perhaps mostly for discoverability, but I have no strong opinion here.

ekharkunov replied:
I thought about making it as a part of constructor. But I refused that idea because in other cases (matrix4 from scale, quaternion from matrix4, matrix4 from translation) also should be done as part of constructor.
Maybe it can be generalized like
```lua
local m = vmath.matrix4({
    translation = vmath.vector3(),
    rotation = vmath.quat(1,1,1,0)
    scale = 4
}}

local m2 = vmath.matrix4(m)
local m3 = vmath.matrix4(vmath.vector({ .... })
```
it's a matter of had several but small and clear function vs have 1 function and hide all implementation under the hood.

britzl replied:
It's not great to have to pass a table every time. It will generate quite a bit of garbage (or the user has to keep a table around only for this purpose)

JCash commented on engine/script/src/script_vmath.cpp:
```diff
@@ -2299,6 +2299,166 @@ namespace dmScript
+        }
+        else
+        {
+            const ScriptUserType argument_type = GetType(L, 1);
+            if (argument_type == SCRIPT_TYPE_VECTOR3)
+            {
```
This is a good scenario to use ToVector3 et al, and then early out.

```
if number:
  do stuff;
  return 1
  
 Vector3* pvec3 = ToVector3(L, 1)
 if pvec3:
    ...
    return 1
 
 return luaL_error();
 ```

JCash commented on engine/script/src/script_vmath.cpp:
```diff
@@ -2299,6 +2299,166 @@ namespace dmScript
+    static int Matrix4_Scale(lua_State* L)
+    {
+        Vector3* scale = CheckVector3(L, 1);
+        Matrix4 result = Matrix4::scale(*scale);
```
I think having a 1/3 number(s) is another use case, as we can then skip necessarily creating a temporary vector3.

JCash commented on engine/script/src/script_vmath.cpp:
```diff
@@ -2299,6 +2299,166 @@ namespace dmScript
+     */
+    static int Vector_Clamp(lua_State* L)
+    {
+        float min = (float) luaL_checknumber(L, 2);
+        float max = (float) luaL_checknumber(L, 3);
```
Potentially, we could also have a "per element" clamp.
I'm not sure what the actual user requirements are here tbh.
We can probably leave it until someone requests it.

ekharkunov replied:
In that case I would suggest to have two separate functions:
* `vmath.clamp(v, min, max)` for number only
* `vmath.vclamp(vec, min_vec, max_vec)` for vectors only

WDYT?

britzl replied:
I prefer not to complicate it. vmath.clamp() for numbers and the elements of vectors will be very useful.

JCash reviewed (changes requested):
Looks good.
Minor name change for consistency. `vmath.quat_from_matrix4`

Also added Björn for api review.

ekharkunov commented on engine/script/src/script_vmath.cpp:
```diff
@@ -2299,6 +2299,166 @@ namespace dmScript
+     * @param matrix [type:matrix4] source matrix4
+     * @return q [type:quaternion] new quaternion
+     */
+    static int Quat_Matrix4(lua_State* L)
+    {
+        Matrix4* matrix = CheckMatrix4(L, 1);
+        PushQuat(L, dmVMath::Quat(matrix->getUpper3x3()));
+        return 1;
+    }
```
In that case I would suggest to rename that function also https://github.com/defold/defold/blob/11f05ad81aa18e68e3cc59cc26a9fc2886935508/engine/script/src/script_vmath.cpp#L2484C11-L2484C30
from `vmath.matrix4_translation` to `vmath.matrix4_from_translation`. Left old one for backward compatability and mark as deprecated.
And that one also rename https://github.com/defold/defold/blob/11f05ad81aa18e68e3cc59cc26a9fc2886935508/engine/script/src/script_vmath.cpp#L2500 to `vmath.matrix4_from_scale`.
WDYT?

britzl replied:
The odd one out is Matrix_FromQuat():

```
{"matrix4_from_quat", Matrix4_FromQuat},
{"matrix4_axis_angle", Matrix4_AxisAngle},
{"matrix4_rotation_x", Matrix4_RotationX},
{"matrix4_rotation_y", Matrix4_RotationY},
{"matrix4_rotation_z", Matrix4_RotationZ},
{"matrix4_translation", Matrix4_Translation},
```

All of the others also create a matrix4 from something but none of them have "from" in the name.

ekharkunov replied:
So rename `matrix4_from_quat` to `matrix4_quat` and left `quat_matrix4` as-is?

britzl replied:
Yes, but keep matri4x_from_quat but remove it from the documentation

ekharkunov committed:
Review fixes.
Introduce new function that returns userdata type and pointer to userdata. For vector3/vector4/quat/matrix4 do additional checks.

ekharkunov commented on engine/script/src/script_vmath.cpp:
```diff
@@ -2299,6 +2299,166 @@ namespace dmScript
+
+        if (translation_type == SCRIPT_TYPE_VECTOR3)
+        {
+            Vector3* translation = CheckVector3(L, 1);
```
@JCash  Suggest following function
https://github.com/defold/defold/pull/9438/files#diff-15a7f3705bd4b3e12269cedf985da8656b010506cf18223a5ea6154253e096b0R138-R192

I've used it across all vmath functions that has variation across input arguments.

ekharkunov committed:
Add vmath.clamp for vector arguments.

ekharkunov committed:
Fix dereferencing issue.

ekharkunov committed:
Add matrix4_scale implementation for 1/3 numbers.

JCash commented on engine/script/src/script_vmath.cpp:
```diff
@@ -2299,6 +2401,293 @@ namespace dmScript
+        {
+            float x, y, z;
+            x = (float) luaL_checknumber(L, 1);
+            if (lua_isnumber(L, 2) && lua_isnumber(L, 3))
+            {
+                y = (float) luaL_checknumber(L, 2);
+                z = (float) luaL_checknumber(L, 3);
+            }
```
I think that if `top > 1` then we can use "check number" regardless?
We don't want to pass in other types, do we?

ekharkunov replied:
> I think that if `top > 1` then we can use "check number" regardless? We don't want to pass in other types, do we?

I think only vector 3 and set of numbers. Nothing else.

ekharkunov replied:
> Another option is to always have 3 numbers?

You can pass 1 or 3 numbers. In case if only 1 number passed - it used as value for every scale component (x, y, z).

JCash commented on engine/script/src/script_vmath.cpp:
````diff
@@ -2299,6 +2401,293 @@ namespace dmScript
+        return 1;
+    }
+
+    /*# creates a new matrix4 from scale vector
+     *
+     * Creates a new matrix constructed from scale vector
+     *
+     * @name vmath.matrix4_scale
+     * @param scale [type:vector3] scale
+     * @return matrix [type:matrix4] new matrix4
+     * @examples
+     *
+     * ```lua
+     * local scale = vmath.vector3(1, 0.5, 0.5)
+     * local result = vmath.matrix4_scale(scale)
+     * print(result) --> vmath.matrix4(1, 0, 0, 0, 0, 0.5, 0, 0, 0, 0, 0.5, 0, 0, 0, 0, 1)
+     * ```
+     */
````
This is a new thing for us I think, having explicit documentation entries for the same function.
I think it's borderline, as it might be tricky to keep the documentation in sync that way. The alternative may be visually awkward (for the arguments) thought.
However, the examples would be grouped together.

Good enough for now at least.

ekharkunov replied:
Not new. Initially I was searching for example how to describe one function with different set of arguments and than found that https://github.com/defold/defold/blob/dev/engine/script/src/script_vmath.cpp#L1064-L1120. And did in the same way.

JCash commented on engine/script/src/script_vmath.cpp:
```diff
@@ -111,11 +111,100 @@ namespace dmScript
+            !isnan(m->getElem(0, 3)) && !isnan(m->getElem(1, 3)) && !isnan(m->getElem(2, 3)) && !isnan(m->getElem(3, 3));
+    }
+
+    static ScriptUserType CheckUserData(lua_State* L, int index, void** userdata)
```
👍

JCash commented on engine/script/src/script_vmath.cpp:
````diff
@@ -2299,6 +2401,293 @@ namespace dmScript
+     * @example
+     * 
+     * ```lua
+     * local value1 = 56
````
Indentation

JCash commented on engine/script/src/script_vmath.cpp:
```diff
@@ -2299,6 +2401,293 @@ namespace dmScript
+                dmMath::Clamp(value->getW(), min_v.getW(), max_v.getW())
+            ));
+        }
+        else if  (argument_type == SCRIPT_TYPE_VECTOR)
```
While I do like the implementation, I also wonder about the use case.
I rarely see anyone using the Vector type, and I think I'd personally just wait until someone needs it before adding it.

We use the mantra "every bit counts", in terms of keeping the engine size as small as possible. Sometimes the change it warranted, and sometime we want to wait until someone asks for it. I believe this case might be the latter.
I don't have an overly strong opinion here.

@britzl what do you think?

britzl replied:
I agree. The Vector type is very rarely used, and if it is used it is likely only to describe an animation curve. I think we can remove it and add a note in the documentation about which types are not supported.

ekharkunov replied:
I've removed `clamp` implementation for `vector` type but left check for `SCRIPT_TYPE_VECTOR` in `CheckUserData` to cover all types defined in `ScriptUserType`.

JCash commented on engine/script/src/script_vmath.cpp:
```diff
@@ -2409,7 +2803,7 @@ namespace dmScript
     {
         Vector4* v = (Vector4*)CheckUserType(L, index, TYPE_HASHES[SCRIPT_TYPE_VECTOR4], 0);
-        if (isnan(v->getX()) || isnan(v->getY()) || isnan(v->getZ()) || isnan(v->getW()))
+        if (!CheckVector4Components(v))
```
👍

JCash commented on engine/script/src/test/test_script_vmath.cpp:
```diff
@@ -292,6 +294,10 @@ TEST_F(ScriptVmathTest, TestMatrix4Fail)
+    ASSERT_FALSE(RunString(L, "local m = vmath.matrix4_compose()"));
 }
 
 
```
I'd like to see some unit test tests for the new clamp function.

JCash reviewed (changes requested):
Questions regarding:
* do we want to support matrix_scale(1) ?
* do we want to support clamp() for vector() type ?
* missing unit tests for clamp

ekharkunov committed:
Remove clamp implementation for vector type.
Fix identation.
Add more clamp tests.
Optimize detection argument count in matrix4_scale function.

JCash commented on engine/script/src/test/test_script_vmath.lua:
```diff
@@ -0,0 +1,54 @@
+-- Copyright 2020-2024 The Defold Foundation
```
👍
